topology.go

```diff
@@ -24,123 +23,122 @@ type TopologyOptions struct {
-            INNER JOIN components AS parent ON component_relationships.relationship_id = parent.id %s
-        )
-    )
-	SELECT json_agg(
-        jsonb_set_lax(
-            jsonb_set_lax(
-                jsonb_set_lax(
-                    to_jsonb(topology_result),
-                        '{checks}', %s
-                ), '{summary,insights}', %s
-            ), '{summary,incidents}', %s
-        )
-    ) :: jsonb FROM topology_result`,
-		opts.componentWhereClause(), opts.componentRelationWhereClause(), opts.checksForComponents(),
-		opts.configAnalysisSummaryForComponents(), opts.incidentSummaryForComponents())
-
+	query := `
+SELECT
+  id,
+  system_template_id,
+  external_id,
+  parent_id,
+  name,
+  text,
+  topology_type,
+  namespace,
+  labels,
+  hidden,
+  silenced,
+  status,
+  COALESCE(description, ''),
+  lifecycle,
+  tooltip,
+  status_reason,
+  schedule,
+  icon,
+  type,
+  owner,
+	configs,
+  path,
+  summary,
+  is_leaf,
+  created_by,
+  created_at,
+  updated_at,
+  deleted_at,
+  log_selectors,
+  checks
+FROM
+  topology`
+
+	var clauses []string
 	args := make(map[string]any)
 	if opts.ID != "" {
+		clauses = append(clauses, "(id = @id OR parent_id = @id)")
 		args["id"] = opts.ID
 	}
+
 	if opts.Owner != "" {
+		clauses = append(clauses, "owner = @owner")
 		args["owner"] = opts.Owner
 	}
-	if opts.Labels != nil {
+
+	if len(opts.Labels) > 0 {
+		clauses = append(clauses, "labels @> @labels")
 		args["labels"] = opts.Labels
 	}
-	return query, args
-}
 
-func (opts TopologyOptions) checksForComponents() string {
-	return `(
-        SELECT json_agg(checks) FROM checks
-        LEFT JOIN check_component_relationships ON checks.id = check_component_relationships.check_id
-        WHERE check_component_relationships.component_id = topology_result.id AND check_component_relationships.deleted_at IS NULL
-        GROUP BY check_component_relationships.component_id
-    ) :: jsonb`
-}
+	if len(clauses) > 0 {
+		query = fmt.Sprintf("%s WHERE %s", query, strings.Join(clauses, " AND "))
+	}
 
-func (opts TopologyOptions) configAnalysisSummaryForComponents() string {
-	return `(SELECT analysis FROM analysis_summary_by_component WHERE id = topology_result.id)`
+	return query, args
 }
 
-func (p TopologyOptions) incidentSummaryForComponents() string {
-	return `(SELECT incidents FROM incident_summary_by_component WHERE id = topology_result.id)`
+func scanComponent(row pgx.Rows) (*models.Component, error) {
+	var c models.Component
+	err := row.Scan(
+		&c.ID,
+		&c.SystemTemplateID,
+		&c.ExternalId,
+		&c.ParentId,
+		&c.Name,
+		&c.Text,
+		&c.TopologyType,
+		&c.Namespace,
+		&c.Labels,
+		&c.Hidden,
+		&c.Silenced,
+		&c.Status,
+		&c.Description,
+		&c.Lifecycle,
+		&c.Tooltip,
+		&c.StatusReason,
+		&c.Schedule,
+		&c.Icon,
+		&c.Type,
+		&c.Owner,
+		&c.Configs,
+		&c.Path,
+		&c.Summary,
+		&c.IsLeaf,
+		&c.CreatedBy,
+		&c.CreatedAt,
+		&c.UpdatedAt,
+		&c.DeletedAt,
+		&c.LogSelectors,
+		&c.Checks,
+	)
```

Comment:
https://pkg.go.dev/github.com/jackc/pgx/v5#RowToStructByPos

Prefer this or a gorm struct

If there are a lot of columns with index being important, it may lead to off by one or other indexing mistakes

Reply:
I'm not really seeing the benefit with `RowToStructByPos`. It looks like you still need to maintain the order in the newly defined struct, no?